### ISSUE TITLE
[Core|Tune] Not all workers cleaned up on Windows on trial failure.

### ISSUE THREAD

**Issue by jbedorf** [ray-project/ray#29388]
### What happened + What you expected to happen

When using Tune (+RLlib) on Windows machines workers are not always properly cleaned up when an actor fails. This results in a slow but steady increase of worker processes and associated memory usage. The log files of the workers indicate that they exit (see screenshot below), but they are actually left running. This does not happen for every worker, and not every time, it seems to happen more on machines with a larger amount of cores (32+). 

The reproduction script enforces this by launching an RLlib run where one of the actors will crash. This will then trigger a relaunch of the run by Tune. This then launches new workers. However, not all the workers are cleaned up. Running the same script on Linux does not show this behaviour, e.g. there is no increase of processes there.

The following screenshot shows this behavior. It shows the active Python processes, sorted by launch time, at the top we have the reproduction script. Followed by the processes launched by `ray.init` and then followed by various workers. This script only launches 5 workers (learner+4 samplers), but as visible here after 5 minutes we're at a couple of dozen active Python processes. 

If we then look into the log file of the worker process with PID `174840` it shows that it should have been killed at 9:00:08, however at 0:04:45 it is still active and taking up nearly 1GB of memory. Note that not just the `default_worker` processes stay active, but also the `setup_worker` scripts. 

Overview of active, related, processes. The one we look into in more detail is the 11th from the top.
![image](https://user-images.githubusercontent.com/1643141/196029569-d1df186b-340d-4b02-b77d-6fcaf6b4f505.png)

Screenshot of the log of the worker with PID 174840.
![image](https://user-images.githubusercontent.com/1643141/196029610-06e99d9e-705d-41ac-94b2-e1baad23218e.png)

### Versions / Dependencies

Python 3.8. Ray 2.0.0 and Ray nightly (per Oct 16). Windows 10.

### Reproduction script

Example script that keeps attempting to complete a run, which will keep failing, thereby slowly causing an increase in process usage.

```python
import gym
import ray
from ray import tune
from ray.rllib.algorithms import ppo
import time

class MyEnvironment(gym.Env):
    def __init__(self, config):
        self.config = config
        self.action_space = gym.spaces.Discrete(1)
        self.observation_space = gym.spaces.Box(low=-1, high=1, shape=(5,))

    def step(self, action):
        time.sleep(5)
        if self.config.worker_index == 2:
            #time.sleep(60)
            raise ValueError("Failing worker")
        return self.observation_space.sample(), 0.0, False, {}

    def reset(self):
        return self.observation_space.sample()


class MyTrainable(tune.Trainable):
    def setup(self, config):
        self.agent = ppo.PPO(config, env=MyEnvironment)

    def step(self):
        return self.agent.train()


ray.init(num_cpus=6)

num_workers = 4
bundles = [{"CPU": 1}] + [{"CPU": 1}] * num_workers
analysis = tune.run(
    MyTrainable,
    stop={"training_iteration": 10},
    max_failures=100,
    config={"num_workers": num_workers, "disable_env_checking": True},
    resources_per_trial=tune.PlacementGroupFactory(bundles),
)

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by mattip** [comment]
I tried to reproduce this, I ended up installing tensorflow which picked up  `tensorflow_intel-2.11.0-cp38-cp38-win_amd64.whl` and downgraded protobuf to ` protobuf-3.19.6`. I also needed to use numpy 1.23.5 to avoid #31293, and "setuptools<60" to avoid an annoying deprecation warning:
```
conda create -y -n issue29388 python=3.8
conda activate issue29388
pip install "ray[rllib]==2.0.0"
pip install tensorflow "numpy==1.23.5" "setuptools<60"
python d:\temp\issue29388.py
```

But then I fall afoul of #27646 which kills the script after only a few minutes. So I [copied the fix](https://raw.githubusercontent.com/mattip/ray/d196291ad7b425fc687f377b086498763ee0b7bf/python/ray/_private/log_monitor.py) from #30643. Now the script ran for a while, killing workers as it progressed. I grepped the logs for the "Force kill" line in the logs, and also wrote out the PIDs of all the active python processes:
```
>findstr "Forcekill " %TEMP%\ray\session_latest\logs\* > d:\temp\killed.txt
>wmic process where caption="python.exe" get processid > d:\temp\active.txt
```

and found I could reproduce the issue. I can attach the visual studio debugger to the process, and "break all" to stop the process. I see a single thread stopped inside `_raylet.pyd` but could not examine anything beyond that: registers could not be displayed. When the trial ended (after 101 failed tests), the process died with all the other ray processes.

**Comment by mattip** [comment]
I checked out the source of 2.0.0 in a new environment and built with `cd python && bazel build  -c dbg //:ray-pkg`. Then I copied the `_raylet.pyd` and `_raylet.pdb` over the ones in the conda environment, and reran the test script in the modified `issue29388` environment, this time I set `max_failures=1000`. Even after 322 rounds, I could not find a crashed process that was still alive, so the problem must be something in the `_raylet.pyd` module.

**Comment by mattip** [comment]
I did notice that `pip install` ing tensorflow, needed for tune, has the side-effect of downgrading protobuf to 3.19.6. xref #28754. I think the original reason to pin protobuf was rather esoteric, and had to do with using older model files. In any case it is doubtful this is causing windows processes to hang on exit.

**Comment by mattip** [comment]
Running this workflow on HEAD, I get an error about ` ValueError: 'observation_space' not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no 'observation_space' specified in config!` but I see an `observation_space` in the `gym.Env`.

<details><summary>Full traceback</summary>

```
d:\pypy_stuff\ray_dev>python \temp\issue29388.py
lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
2023-02-28 17:46:24,459 INFO worker.py:1557 -- Started a local Ray instance.
(pid=672) 2023-02-28 17:46:33,709       WARNING compression.py:16 -- lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
(MyTrainable pid=672) 2023-02-28 17:46:34,506   WARNING deprecation.py:50 -- DeprecationWarning: `algo = Algorithm(env='<class '__main__.MyEnvironment'>', ...)` has been deprecated. Use `algo = AlgorithmConfig().environment('<class '__main__.MyEnvironment'>').build()` instead. This will raise an error in the future!
(MyTrainable pid=672) 2023-02-28 17:46:34,506   INFO algorithm_config.py:2981 -- Your framework setting is 'tf', meaning you are using static-graph mode. Set framework='tf2' to enable eager execution with tf2.x. You may also then want to set eager_tracing=True in order to reach similar execution speed as with static-graph mode.
(MyTrainable pid=672) 2023-02-28 17:46:34,568   INFO algorithm.py:501 -- Current log_level is WARN. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
(pid=3372) 2023-02-28 17:46:41,552      WARNING compression.py:16 -- lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
(pid=9824) 2023-02-28 17:46:41,552      WARNING compression.py:16 -- lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
(pid=3272) 2023-02-28 17:46:41,568      WARNING compression.py:16 -- lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
(RolloutWorker pid=3372) 2023-02-28 17:46:41,928        ERROR worker.py:772 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=3372, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000018AE7998B20>)
(RolloutWorker pid=3372)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(RolloutWorker pid=3372)     raise e
(RolloutWorker pid=3372)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(RolloutWorker pid=3372)     with ray._private.worker._changeproctitle(title, next_title):
(RolloutWorker pid=3372)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(RolloutWorker pid=3372)     outputs = function_executor(*args, **kwargs)
(RolloutWorker pid=3372)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(RolloutWorker pid=3372)     return function(actor, *arguments, **kwarguments)
(RolloutWorker pid=3372)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(RolloutWorker pid=3372)     return method(__ray_actor, *args, **kwargs)
(RolloutWorker pid=3372)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(RolloutWorker pid=3372)     return method(self, *_args, **_kwargs)
(RolloutWorker pid=3372)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(RolloutWorker pid=3372)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(RolloutWorker pid=3372)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(RolloutWorker pid=3372)     raise ValueError(
(RolloutWorker pid=3372) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(RolloutWorker pid=9824) 2023-02-28 17:46:41,928        ERROR worker.py:772 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=9824, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x00000227570E7BB0>)
(RolloutWorker pid=9824)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(RolloutWorker pid=9824)     raise e
(RolloutWorker pid=9824)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(RolloutWorker pid=9824)     with ray._private.worker._changeproctitle(title, next_title):
(RolloutWorker pid=9824)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(RolloutWorker pid=9824)     outputs = function_executor(*args, **kwargs)
(RolloutWorker pid=9824)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(RolloutWorker pid=9824)     return function(actor, *arguments, **kwarguments)
(RolloutWorker pid=9824)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(RolloutWorker pid=9824)     return method(__ray_actor, *args, **kwargs)
(RolloutWorker pid=9824)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(RolloutWorker pid=9824)     return method(self, *_args, **_kwargs)
(RolloutWorker pid=9824)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(RolloutWorker pid=9824)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(RolloutWorker pid=9824)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(RolloutWorker pid=9824)     raise ValueError(
(RolloutWorker pid=9824) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(RolloutWorker pid=3272) 2023-02-28 17:46:41,928        ERROR worker.py:772 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=3272, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000023CAAC39B20>)
(RolloutWorker pid=3272)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(RolloutWorker pid=3272)     raise e
(RolloutWorker pid=3272)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(RolloutWorker pid=3272)     with ray._private.worker._changeproctitle(title, next_title):
(RolloutWorker pid=3272)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(RolloutWorker pid=3272)     outputs = function_executor(*args, **kwargs)
(RolloutWorker pid=3272)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(RolloutWorker pid=3272)     return function(actor, *arguments, **kwarguments)
(RolloutWorker pid=3272)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(RolloutWorker pid=3272)     return method(__ray_actor, *args, **kwargs)
(RolloutWorker pid=3272)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(RolloutWorker pid=3272)     return method(self, *_args, **_kwargs)
(RolloutWorker pid=3272)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(RolloutWorker pid=3272)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(RolloutWorker pid=3272)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(RolloutWorker pid=3272)     raise ValueError(
(RolloutWorker pid=3272) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(pid=9748) 2023-02-28 17:46:42,662      WARNING compression.py:16 -- lz4 not available, disabling sample compression. This will significantly impact RLlib performance. To install lz4, run `pip install lz4`.
(MyTrainable pid=672) 2023-02-28 17:46:43,065   ERROR actor_manager.py:507 -- Ray error, taking actor 1 out of service. The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=3372, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000018AE7998B20>)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(MyTrainable pid=672)     raise e
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(MyTrainable pid=672)     with ray._private.worker._changeproctitle(title, next_title):
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(MyTrainable pid=672)     outputs = function_executor(*args, **kwargs)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(MyTrainable pid=672)     return function(actor, *arguments, **kwarguments)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(MyTrainable pid=672)     return method(__ray_actor, *args, **kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(MyTrainable pid=672)     return method(self, *_args, **_kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(MyTrainable pid=672)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(MyTrainable pid=672)     raise ValueError(
(MyTrainable pid=672) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(MyTrainable pid=672) 2023-02-28 17:46:43,065   ERROR actor_manager.py:507 -- Ray error, taking actor 2 out of service. The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=9748, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001394E138B50>)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(MyTrainable pid=672)     raise e
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(MyTrainable pid=672)     with ray._private.worker._changeproctitle(title, next_title):
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(MyTrainable pid=672)     outputs = function_executor(*args, **kwargs)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(MyTrainable pid=672)     return function(actor, *arguments, **kwarguments)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(MyTrainable pid=672)     return method(__ray_actor, *args, **kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(MyTrainable pid=672)     return method(self, *_args, **_kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(MyTrainable pid=672)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(MyTrainable pid=672)     raise ValueError(
(MyTrainable pid=672) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(MyTrainable pid=672) 2023-02-28 17:46:43,067   ERROR actor_manager.py:507 -- Ray error, taking actor 3 out of service. The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=3272, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000023CAAC39B20>)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(MyTrainable pid=672)     raise e
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(MyTrainable pid=672)     with ray._private.worker._changeproctitle(title, next_title):
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(MyTrainable pid=672)     outputs = function_executor(*args, **kwargs)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(MyTrainable pid=672)     return function(actor, *arguments, **kwarguments)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(MyTrainable pid=672)     return method(__ray_actor, *args, **kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(MyTrainable pid=672)     return method(self, *_args, **_kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(MyTrainable pid=672)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(MyTrainable pid=672)     raise ValueError(
(MyTrainable pid=672) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(MyTrainable pid=672) 2023-02-28 17:46:43,068   ERROR actor_manager.py:507 -- Ray error, taking actor 4 out of service. The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=9824, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x00000227570E7BB0>)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(MyTrainable pid=672)     raise e
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(MyTrainable pid=672)     with ray._private.worker._changeproctitle(title, next_title):
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(MyTrainable pid=672)     outputs = function_executor(*args, **kwargs)
(MyTrainable pid=672)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(MyTrainable pid=672)     return function(actor, *arguments, **kwarguments)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(MyTrainable pid=672)     return method(__ray_actor, *args, **kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(MyTrainable pid=672)     return method(self, *_args, **_kwargs)
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(MyTrainable pid=672)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(MyTrainable pid=672)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(MyTrainable pid=672)     raise ValueError(
(MyTrainable pid=672) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
(RolloutWorker pid=9748) 2023-02-28 17:46:43,044        ERROR worker.py:772 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::RolloutWorker.__init__() (pid=9748, ip=127.0.0.1, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001394E138B50>)
(RolloutWorker pid=9748)   File "python\ray\_raylet.pyx", line 910, in ray._raylet.execute_task
(RolloutWorker pid=9748)     raise e
(RolloutWorker pid=9748)   File "python\ray\_raylet.pyx", line 865, in ray._raylet.execute_task
(RolloutWorker pid=9748)     with ray._private.worker._changeproctitle(title, next_title):
(RolloutWorker pid=9748)   File "python\ray\_raylet.pyx", line 869, in ray._raylet.execute_task
(RolloutWorker pid=9748)     outputs = function_executor(*args, **kwargs)
(RolloutWorker pid=9748)   File "python\ray\_raylet.pyx", line 811, in ray._raylet.execute_task.function_executor
(RolloutWorker pid=9748)     return function(actor, *arguments, **kwarguments)
(RolloutWorker pid=9748)   File "d:\pypy_stuff\ray_dev\python\ray\_private\function_manager.py", line 674, in actor_method_executor
(RolloutWorker pid=9748)     return method(__ray_actor, *args, **kwargs)
(RolloutWorker pid=9748)   File "d:\pypy_stuff\ray_dev\python\ray\util\tracing\tracing_helper.py", line 466, in _resume_span
(RolloutWorker pid=9748)     return method(self, *_args, **_kwargs)
(RolloutWorker pid=9748)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\evaluation\rollout_worker.py", line 687, in __init__
(RolloutWorker pid=9748)     self.policy_dict, self.is_policy_to_train = self.config.get_multi_agent_setup(
(RolloutWorker pid=9748)   File "d:\pypy_stuff\ray_dev\python\ray\rllib\algorithms\algorithm_config.py", line 2530, in get_multi_agent_setup
(RolloutWorker pid=9748)     raise ValueError(
(RolloutWorker pid=9748) ValueError: `observation_space` not provided in PolicySpec for default_policy and env does not have an observation space OR no spaces received from other workers' env(s) OR no `observation_space` specified in config!
```

</details>

**Comment by jbedorf** [comment]
Leaving a comment to keep this issue open as, as far as I know, it is still occurring.

**Comment by anyscalesam** [comment]
@mattip can you repro on latest tf + ray28 and than we can from there.

**Comment by mattip** [comment]
I ran this to reproduce:
```
conda create -y -n issue29388 python=3.10
conda activate issue29388
conda install -y pip
pip install "ray[rllib]==2.8.0 gputil
pip install tensorflow torch 
python d:\temp\issue29388.py
```

It installed `tensorflow_intel-2.14.0`, which brought with it `grpcio-1.59.2`. The original reproducer used `gym`, I replaced that with `import gymnasium as gym`. I also added `from gymnasium.wrappers import EnvCompatibility` and changed the `gym.Env` base class to  `EnvCompatibility`. Now the script runs, but I do not see the `Force kill` text in the logs.

@jbedorf could you update the reproducer and see if the issue still occurs with latest software versions?

**Comment by jbedorf** [comment]
Give me a week to get this set up and working again and reproduce the test.

**Comment by anyscalesam** [comment]
@jbedorf any luck?

**Comment by jbedorf** [comment]
No luck (which is a good thing in this case).
I've done various attempts at reproducing the issue using the latest Ray version, but after running the same setup as original I was unable to get any leaked processes. I did see the 'force kill' events in the logs, but those processes were properly killed in this case and there were no dangling processes like last time. Ran the test for a couple of hours to verify. So looks like we can close this issue. Thanks!
